h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -164,76 +166,142 @@ public AutoML() {
+      _runCountdown = Countdown.fromSeconds(_buildSpec.build_control.stopping_criteria.max_runtime_secs());
 
+      prepareData();
+      initLeaderboard();
       planWork();
-
       _modelingStepsExecutor = new ModelingStepsExecutor(_leaderboard, _eventLog, _runCountdown);
     } catch (Exception e) {
       delete(); //cleanup potentially leaked keys
       throw e;
     }
   }
 
-  private void initLeaderboard(AutoMLBuildSpec buildSpec) {
-    String sort_metric = buildSpec.input_spec.sort_metric;
-    sort_metric = sort_metric == null || StoppingMetric.AUTO.name().equalsIgnoreCase(sort_metric) ? null : sort_metric.toLowerCase();
-    _leaderboard = Leaderboard.getOrMake(_key.toString(), _eventLog, _leaderboardFrame, sort_metric);
-  }
+  private void validateBuildSpec(AutoMLBuildSpec buildSpec) {
+    // Validate all buildSpec parameters here and provide reasonable defaults or parameter cleaning if necessary.
+    // When assigning default or modifying the original value, use `eventLog` to inform the user about the changes.
+    // Please no additional logic!
 
-  private void handleReproducibilityParameters(AutoMLBuildSpec buildSpec) {
-    eventLog().info(Stage.Workflow, "Build control seed: " + buildSpec.build_control.stopping_criteria.seed() +
-            (buildSpec.build_control.stopping_criteria.seed() == -1 ? " (random)" : ""));
+    // Can possibly use this validation entry point to provide parameters validation for Flow later.
+    // Can also reuse ModelBuilder.ValidationMessage for this.
+
+    validateInput(buildSpec.input_spec);
+    validateModelValidation(buildSpec);
+    validateModelBuilding(buildSpec.build_models);
+    validateEarlyStopping(buildSpec.build_control.stopping_criteria, buildSpec.input_spec);
+    validateReproducibility(buildSpec);
   }
 
-  private void handleEarlyStoppingParameters(AutoMLBuildSpec buildSpec) {
-    if (buildSpec.build_control.stopping_criteria.stopping_tolerance() == AUTO_STOPPING_TOLERANCE) {
-      buildSpec.build_control.stopping_criteria.set_default_stopping_tolerance_for_frame(_trainingFrame);
-      eventLog().info(Stage.Workflow, "Setting stopping tolerance adaptively based on the training frame: " +
-              buildSpec.build_control.stopping_criteria.stopping_tolerance());
-    } else {
-      eventLog().info(Stage.Workflow, "Stopping tolerance set by the user: " + buildSpec.build_control.stopping_criteria.stopping_tolerance());
-      double default_tolerance = AutoMLBuildSpec.AutoMLStoppingCriteria.default_stopping_tolerance_for_frame(_trainingFrame);
-      if (buildSpec.build_control.stopping_criteria.stopping_tolerance() < 0.7 * default_tolerance){
-        eventLog().warn(Stage.Workflow, "Stopping tolerance set by the user is < 70% of the recommended default of " + default_tolerance + ", so models may take a long time to converge or may not converge at all.");
+  private void validateInput(AutoMLInput input) {
+    if (null == DKV.getGet(input.training_frame))
```

Comment:
@seb-h2o minor. this is not a Java style to put null upfront I believe

Reply:
I've always seen both. 
I must have copied it because I usually don't use it this way, but I can change if it hurts your eyes.